examples/user_defined_iface.yml

```diff
@@ -0,0 +1,43 @@
+---
+user-defined:
+    interface-types:
+        - name: vxcan
+          handler-script: |
+            #!/bin/bash
+            ifname=$1
+            action=$2
+
+            if [ "$action" = "device-add" ]; then
+                peer=$CONNECTION_USER_VXCAN__PEER
+
+                if [ -z "$peer" ]; then
+                    echo "ERROR=Missing peer name"
+                    exit 1
+                fi
+
+                if ! err=$(ip link add "$ifname" type vxcan peer "$peer" 2>&1);
+                then
+                    echo "ERROR=Failed creating the interface: $err"
+                    exit 2
+                fi
+                ip link set $peer up
+
+                echo IFINDEX="$(cat /sys/class/net/"$ifname"/ifindex)"
+                exit 0
+            elif [ "$action" = "device-delete" ]; then
+                # Delete the interface created by "device-add" here.
+                ip link del "$ifname"
+                exit 0
+            fi
+
+interfaces:
+    - name: vxcan1
+      type: vxcan
```

Comment:
IMHO user-defined types need their own namespace, otherwise we will have a hard time in the future to add new internal interface types. Maybe something like: `type: generic: vxcan`.

Reply:
`generic` is not explicit enough, many thing can fit into it. I think `user-defined` is clear enough without any document explain what it is.

Reply:
My concern here is that `type: vxcan` needs to have it's own namespace because this way user defined types will collide with possible type we will define in the the future in core Nmstate. So with `user-defined`, it would be something like `type: user-defined: vxcan`. I am more in favor of `type: plugin: vxcan` now. Another option would be to use `type: plugin` and move all plugin details to `plugin-options`. The type can then be optional and the presence of `plugin-options` could be enough to identify this as type `plugin`.:

`name: vxcan1
type: plugin
 plugin-options:
    type: vxcan
    state:
        peer: "vxcan1-ep"
`